common/mqtt_tests/src/test_mqtt_client.rs

```diff
@@ -0,0 +1,170 @@
+use crate::with_timeout::WithTimeout;
+use rumqttc::{AsyncClient, Event, EventLoop, MqttOptions, Packet, QoS};
+use std::time::Duration;
+use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};
+
+/// Returns the stream of messages received on a specific topic.
+///
+/// To ease testing, the errors are returned as messages.
+pub async fn messages_published_on(mqtt_port: u16, topic: &str) -> UnboundedReceiver<String> {
+    let (sender, recv) = tokio::sync::mpsc::unbounded_channel();
+
+    let mut con = TestCon::new(mqtt_port);
+
+    if let Err(err) = con.subscribe(topic, QoS::AtLeastOnce).await {
+        let msg = format!("Error: {:?}", err).to_string();
+        let _ = sender.send(msg);
+        return recv;
+    }
+
+    tokio::spawn(async move {
+        con.forward_received_messages(sender).await;
+    });
+
+    recv
+}
+
+/// Check that a list of messages has been received in the given order
+pub async fn assert_received<T>(
+    messages: &mut UnboundedReceiver<String>,
+    timeout: Duration,
+    expected: T,
+) where
+    T: IntoIterator,
+    T::Item: ToString,
+{
+    for expected_msg in expected.into_iter() {
+        let actual_msg = messages.recv().with_timeout(timeout).await;
+        assert_eq!(actual_msg, Ok(Some(expected_msg.to_string())));
+    }
+}
+
+/// Publish a message
+///
+/// Return only when the message has been acknowledged.
+pub async fn publish(mqtt_port: u16, topic: &str, payload: &str) -> Result<(), anyhow::Error> {
```

Comment:
Would it make sense to expose a `subscribe()` API as well similar to this `publish` API, that guarantees a successful subscription? Might help in those tests using random sleeps to make sure that the subscription succeeded.

Reply:
There is such one method `TestCon::subscribe`.